Bottom: 27f6112453ab2f5c2845a6599571ce12e16ef122
Top:    f3d9790cdbe9f8962212f1c7078fefe1719c7c36
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-01-07 23:44:06 +0000

hid/gtk: Fix prelight state for layer button to show current status

One of the ways we identify whether a layer is switched on or off is
by changing the background colour of the GtkToggleButton representing
that layer.

Fix the prelight (mouse-over) background style to match the current
button state, so as to give immediate feed-back when the user presses
the button, and make the interface less confusing.

Closes-bug: lp-699150


---

diff --git a/src/hid/gtk/gui-top-window.c b/src/hid/gtk/gui-top-window.c
index b4c8c42..f2cdb99 100644
--- a/src/hid/gtk/gui-top-window.c
+++ b/src/hid/gtk/gui-top-window.c
@@ -1465,17 +1465,19 @@ layer_enable_button_cb (GtkWidget * widget, gpointer data)
 }
 
 static void
-layer_button_set_color (LayerButtonSet * lb, gchar * color_string)
+layer_button_set_color (LayerButtonSet * lb, gchar * color_string,
+                        bool set_prelight)
 {
   GdkColor color;
 
   if (!lb->layer_enable_ebox)
     return;
-  
+
   color.red = color.green = color.blue = 0;
   ghid_map_color_string (color_string, &color);
   gtk_widget_modify_bg (lb->layer_enable_ebox, GTK_STATE_ACTIVE, &color);
-  gtk_widget_modify_bg (lb->layer_enable_ebox, GTK_STATE_PRELIGHT, &color);
+  gtk_widget_modify_bg (lb->layer_enable_ebox, GTK_STATE_PRELIGHT,
+                        set_prelight ? &color : NULL);
 
   gtk_widget_modify_fg (lb->label, GTK_STATE_ACTIVE, &WhitePixel);
 }
@@ -1573,7 +1575,7 @@ make_layer_buttons (GtkWidget * vbox, GHidPort * port)
       lb->text = g_strdup (text);
       lb->label = label;
 
-      layer_button_set_color (lb, color_string);
+      layer_button_set_color (lb, color_string, active);
       gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (button), active);
 
       g_signal_connect (G_OBJECT (button), "toggled",
@@ -1605,11 +1607,13 @@ ghid_layer_buttons_color_update (void)
 
   for (i = 0; i < N_LAYER_BUTTONS; ++i)
     {
+      bool active;
+
       lb = &layer_buttons[i];
 
       layer_process (&color_string, NULL, NULL, i);
-
-      layer_button_set_color (lb, color_string);
+      active = gtk_toggle_button_get_active (GTK_TOGGLE_BUTTON (lb));
+      layer_button_set_color (lb, color_string, active);
     }
 }
 
@@ -1621,6 +1625,7 @@ ghid_layer_enable_buttons_update (void)
 {
   LayerButtonSet *lb;
   gchar *s;
+  gchar *color_string;
   gint i;
 
 #ifdef DEBUG_MENUS
@@ -1651,6 +1656,8 @@ ghid_layer_enable_buttons_update (void)
 	    printf ("ghid_layer_enable_buttons_update: active=%d new=%d\n",
 		    active, new);
 	}
+      layer_process (&color_string, NULL, NULL, i);
+      layer_button_set_color (lb, color_string, PCB->Data->Layer[i].On);
       gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON
 				    (lb->layer_enable_button),
 				    PCB->Data->Layer[i].On);
